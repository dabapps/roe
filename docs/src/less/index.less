@import '../../../src/less/index.less';

@body-background: @grey-lighter;

h1,
h2 {
  color: @secondary;
}

.container {
  box-shadow: @shadow-hard;
}

.display-columns {
  .column {
    background-color: @grey-lightest;
    border: @border-base;
  }
}
