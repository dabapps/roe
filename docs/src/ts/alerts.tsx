import * as React from 'react';

import {
  Alert,
  CodeBlock,
  Column,
  DabIpsum,
  Row,
  Section
} from '../../../src/ts';

const Alerts = () => (
  <Section>
    <Row>
      <Column>
        <h2>
          Alerts
        </h2>

        <Alert>
          <DabIpsum count={2} />
        </Alert>
        <Alert className="info">
          <DabIpsum count={2} />
        </Alert>
        <Alert className="success">
          <DabIpsum count={2} />
        </Alert>
        <Alert className="warning">
          <DabIpsum count={2} />
        </Alert>
        <Alert className="error">
          <DabIpsum count={2} />
        </Alert>
      </Column>
    </Row>
    <Row>
      <Column>
        <h3>
          Code
        </h3>
        <CodeBlock language="javascript" name="JSX">
          {`
            <Alert>
              <DabIpsum count={2} />
            </Alert>
            <Alert className="info">
              <DabIpsum count={2} />
            </Alert>
            <Alert className="success">
              <DabIpsum count={2} />
            </Alert>
            <Alert className="warning">
              <DabIpsum count={2} />
            </Alert>
            <Alert className="error">
              <DabIpsum count={2} />
            </Alert>
          `}
        </CodeBlock>
      </Column>
    </Row>
  </Section>
);

export default Alerts;
