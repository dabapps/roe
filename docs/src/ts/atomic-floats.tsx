import * as React from 'react';

import {
  CodeBlock,
  Column,
  Row,
  Section
} from '../../../src/ts';

const AtomicFloats = () => (
  <Section>
    <Row>
      <Column>
        <h2>
          Atomic Float Classes
        </h2>
      </Column>
    </Row>
    <Row>
      <Column className="display-columns">
        <h3>
          Demo
        </h3>
        <Row>
          <Column xs={6} className="xs-float-right sm-float-left md-float-right">
            float-right sm-float-left md-float-right
          </Column>
          <Column xs={6} className="xs-float-left sm-float-right md-float-left">
            float-left sm-float-right md-float-left
          </Column>
        </Row>
        <p>
          Note: float-left is an alias for xs-float-left, and float-right an alias for xs-float-right
        </p>
      </Column>
    </Row>
    <Row>
      <Column>
        <h3>
          Code
        </h3>
        <CodeBlock language="javascript">
          {`
            <Row>
              <Column xs={6} className="xs-float-right sm-float-left md-float-right">
                float-right sm-float-left md-float-right
              </Column>
              <Column xs={6} className="xs-float-left sm-float-right md-float-left">
                float-left sm-float-right md-float-left
              </Column>
            </Row>
          `}
        </CodeBlock>
      </Column>
    </Row>
  </Section>
);

export default AtomicFloats;
