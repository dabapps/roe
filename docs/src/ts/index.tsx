import * as React from 'react';
import * as ReactDOM from 'react-dom';

import './livereload';

import {
  Column,
  Container,
  Row
} from '../../../src/ts';

import { Buttons } from './buttons';
import { CodeBlocks } from './code-blocks';
import { Grid } from './grid';
import { Inputs } from './inputs';
import { Text } from './text';

const packageJson = require( '../../../package.json'); // tslint:disable-line:no-var-requires

class App extends React.Component<void, void> {
  public render () {
    return (
      <Container>
        <Row>
          <Column>
            <h1>
              Roe - DabApps' Project Development Kit (v{packageJson.version})
            </h1>
            <p>
              <strong>
                A Collection of React Components for Project Development
              </strong>
            </p>
          </Column>
        </Row>
        <Buttons />
        <CodeBlocks />
        <Grid />
        <Inputs />
        <Text />
      </Container>
    );
  }
}

ReactDOM.render(<App />, document.getElementById('app'));
