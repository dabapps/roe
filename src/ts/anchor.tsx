import * as React from 'react';

const MATCHES_AMPERSAND = /&/gi;
const MATCHES_NON_WORD_CHARACTERS = /[\W_]+/gi;
const MATCHES_LEADING_AND_TRAILING_HYPHENS = /(^-+|-+$)/gi;

export const getHref = (children?: React.ReactNode, href?: string): string | undefined => {
  if (href) {
    return href;
  }

  if (typeof children !== 'string') {
    return undefined;
  }

  return children
    .replace(MATCHES_AMPERSAND, '-and-')
    .replace(MATCHES_NON_WORD_CHARACTERS, '-')
    .replace(MATCHES_LEADING_AND_TRAILING_HYPHENS, '')
    .toLowerCase();
}

export const Anchor: React.SFC<React.HTMLProps<HTMLAnchorElement>> = (props) => {
  const { children, href, ...remainingProps } = props;

  const automaticHref = getHref(children, href);

  return (
    <a {...remainingProps} id={automaticHref} href={automaticHref ? '#' + automaticHref : undefined}>
      {children}
    </a>
  );
};
