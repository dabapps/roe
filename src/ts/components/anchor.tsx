import * as React from 'react';
import { HTMLProps, StatelessComponent } from 'react';
import { getHref } from '../utils';

export type AnchorProps = HTMLProps<HTMLElement>;

export const Anchor: StatelessComponent<AnchorProps> = (props) => {
  const { children, href, ...remainingProps } = props;

  const automaticHref = getHref(children, href);

  return (
    <a {...remainingProps} id={automaticHref} href={automaticHref ? '#' + automaticHref : undefined}>
      {children}
    </a>
  );
}
