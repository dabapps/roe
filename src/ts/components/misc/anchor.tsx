import * as React from 'react';
import { HTMLProps, PureComponent } from 'react';
import { getHref } from '../utils';

export type AnchorProps = HTMLProps<HTMLElement>;

/**
 * Link component that automatically creates an id and hash href that match, for linking to elements on a single page.
 */
export class Anchor extends PureComponent<AnchorProps, {}> {
  public render() {
    const { children, href, ...remainingProps } = this.props;

    const automaticHref = getHref(children, href);

    return (
      <a
        {...remainingProps}
        id={automaticHref}
        href={automaticHref ? '#' + automaticHref : undefined}
      >
        {children}
      </a>
    );
  }
}

export default Anchor;
