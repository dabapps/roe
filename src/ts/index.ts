export { default as Alert } from './components/misc/alert';
export { default as AppRoot } from './components/app/root';
export { default as Anchor } from './components/misc/anchor';
export { default as Badge } from './components/misc/badge';
export { default as Banner } from './components/banners/banner';
export { default as Button } from './components/forms/button';
export { default as CodeBlock } from './components/misc/code-block';
export { default as Collapse } from './components/misc/collapse';
export { default as Column } from './components/grid/column';
export { default as Container } from './components/grid/container';
export { default as ContentBox } from './components/content/content-box';
export { default as ContentBoxHeader } from './components/content/content-box-header';
export { default as ContentBoxFooter } from './components/content/content-box-footer';
export { default as CookieBanner } from './components/banners/cookie-banner';
export { CookieBannerRenderProps } from './components/banners/cookie-banner';
export { default as DabIpsum } from './components/prototyping/dab-ipsum';
export { default as Footer } from './components/navigation/footer';
export { default as FormGroup } from './components/forms/form-group';
export { default as Highlight } from './components/misc/highlight';
export { default as InputGroup } from './components/forms/input-group';
export { default as InputGroupAddon } from './components/forms/input-group-addon';
export { default as InputWithPrefixSuffix } from './components/precomposed/input-with-prefix-suffix';
export { default as Modal } from './components/modals/modal';
export { default as ModalBody } from './components/modals/modal-body';
export { default as ModalCloseIcon } from './components/modals/modal-close-icon';
export { default as ModalFooter } from './components/modals/modal-footer';
export { default as ModalHeader } from './components/modals/modal-header';
export { default as ModalRenderer } from './components/modals/modal-renderer';
export { default as Nav } from './components/navigation/nav';
export { default as NavBar } from './components/navigation/nav-bar';
export { default as NavItem } from './components/navigation/nav-item';
export { default as Row } from './components/grid/row';
export { default as Section } from './components/content/section';
export { default as SideBar } from './components/navigation/side-bar';
export { default as SpacedGroup } from './components/misc/spaced-group';
export { default as SpeechBubble } from './components/misc/speech-bubble';
export { default as Tabs } from './components/tabs/tabs';
export { default as Tab } from './components/tabs/tab';
export { default as Table } from './components/tables/table';
export { default as TableBody } from './components/tables/table-body';
export { default as TableCell } from './components/tables/table-cell';
export { default as TableHead } from './components/tables/table-head';
export { default as TableHeader } from './components/tables/table-header';
export { default as TableRow } from './components/tables/table-row';
export { default as Well } from './components/misc/well';
