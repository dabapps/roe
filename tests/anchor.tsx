import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Anchor } from '../src/ts/';
import { getHref } from '../src/ts/anchor';

describe('Anchor', () => {

  it('should match snapshot', () => {
    const tree = renderer.create(
      <Anchor />
    );

    expect(tree).toMatchSnapshot();
  });

  it('should take regular element attributes', () => {
    const tree = renderer.create(
      <Anchor className="my-class" />
    );

    expect(tree).toMatchSnapshot();
  });

  it('should have an id and href', () => {
    const tree = renderer.create(
      <Anchor href="my-href" />
    );

    expect(tree).toMatchSnapshot();
  });

  describe('getHref', () => {

    it('should return a suitable href', () => {
      expect(getHref()).toBe(undefined);
      expect(getHref('nope', 'yup')).toBe('yup');
      expect(getHref('yup', undefined)).toBe('yup');
      expect(getHref('A regular text string - oh, special chars')).toBe('a-regular-text-string-oh-special-chars');
      expect(getHref(' - Leading and trailing whitespace ,; ')).toBe('leading-and-trailing-whitespace');
    });

  });

});
