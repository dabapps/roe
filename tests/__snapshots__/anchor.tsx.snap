// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor should have an id and href 1`] = `
<a
  href="#my-href"
  id="my-href"
/>
`;

exports[`Anchor should match snapshot 1`] = `<a />`;

exports[`Anchor should take regular element attributes 1`] = `
<a
  className="my-class"
/>
`;
